source/common/upstream/health_checker_impl.h

```diff
@@ -417,5 +421,102 @@ class RedisHealthCheckerImpl : public HealthCheckerImplBase {
+                                        public Http::StreamCallbacks {
+    GrpcActiveHealthCheckSession(GrpcHealthCheckerImpl& parent, HostSharedPtr host);
+    ~GrpcActiveHealthCheckSession();
+
+    void onRpcComplete(Grpc::Status::GrpcStatus grpc_status, const std::string& grpc_message,
+                       bool end_stream);
+    bool isHealthCheckSucceeded(Grpc::Status::GrpcStatus grpc_status) const;
+    void resetState();
+    void logHealthCheckStatus(Grpc::Status::GrpcStatus grpc_status,
+                              const std::string& grpc_message);
+
+    // ActiveHealthCheckSession
+    void onInterval() override;
+    void onTimeout() override;
+
+    // Http::StreamDecoder
+    void decodeHeaders(Http::HeaderMapPtr&& headers, bool end_stream) override;
+    void decodeData(Buffer::Instance&, bool end_stream) override;
+    void decodeTrailers(Http::HeaderMapPtr&&) override;
+
+    // Http::StreamCallbacks
+    void onResetStream(Http::StreamResetReason reason) override;
+    void onAboveWriteBufferHighWatermark() override {}
+    void onBelowWriteBufferLowWatermark() override {}
+
+    void onEvent(Network::ConnectionEvent event);
+    void onGoAway();
+
+    class ConnectionCallbackImpl : public Network::ConnectionCallbacks {
+    public:
+      ConnectionCallbackImpl(GrpcActiveHealthCheckSession& parent) : parent_(parent) {}
+      // Network::ConnectionCallbacks
+      void onEvent(Network::ConnectionEvent event) override { parent_.onEvent(event); }
+      void onAboveWriteBufferHighWatermark() override {}
+      void onBelowWriteBufferLowWatermark() override {}
+
+    private:
+      GrpcActiveHealthCheckSession& parent_;
+    };
+
+    class HttpConnectionCallbackImpl : public Http::ConnectionCallbacks {
+    public:
+      HttpConnectionCallbackImpl(GrpcActiveHealthCheckSession& parent) : parent_(parent) {}
+      // Http::ConnectionCallbacks
+      void onGoAway() override { parent_.onGoAway(); }
+
+    private:
+      GrpcActiveHealthCheckSession& parent_;
+    };
+
+    ConnectionCallbackImpl connection_callback_impl_{*this};
+    HttpConnectionCallbackImpl http_connection_callback_impl_{*this};
+    GrpcHealthCheckerImpl& parent_;
+    Http::CodecClientPtr client_;
+    Http::StreamEncoder* request_encoder_;
+    Grpc::Decoder decoder_;
+    std::unique_ptr<grpc::health::v1::HealthCheckResponse> health_check_response_;
+    // If true, stream reset was initiated by us (GrpcActiveHealthCheckSession), not by HTTP stack,
+    // e.g. remote reset. In this case healthcheck status has already been reported, only state
+    // cleanup is required.
+    bool expect_reset_ = false;
+  };
+
+  virtual Http::CodecClientPtr createCodecClient(Upstream::Host::CreateConnectionData& data) PURE;
+
+  // HealthCheckerImplBase
+  ActiveHealthCheckSessionPtr makeSession(HostSharedPtr host) override {
+    return ActiveHealthCheckSessionPtr{new GrpcActiveHealthCheckSession(*this, host)};
```

Comment:
nit: std::make_unique (please fix elsewhere in this file for similar code, since we are using C++14 now).

Reply:
done